confgenerator/config.go

```diff
@@ -72,65 +148,196 @@ func ParseUnifiedConfigAndValidate(input []byte, platform string) (UnifiedConfig
+type loggingProcessorMap map[string]LoggingProcessor
+type loggingExporterMap map[string]LoggingExporter
 type Logging struct {
-	Receivers  map[string]*LoggingReceiver  `yaml:"receivers,omitempty"`
-	Processors map[string]*LoggingProcessor `yaml:"processors,omitempty"`
-	Exporters  map[string]*LoggingExporter  `yaml:"exporters,omitempty"`
-	Service    *LoggingService              `yaml:"service"`
+	Receivers  loggingReceiverMap  `yaml:"receivers,omitempty" validate:"dive,keys,startsnotwith=lib:"`
+	Processors loggingProcessorMap `yaml:"processors,omitempty" validate:"dive,keys,startsnotwith=lib:"`
+	Exporters  loggingExporterMap  `yaml:"exporters,omitempty"`
+	Service    *LoggingService     `yaml:"service"`
 }
 
-type LoggingReceiverFiles struct {
-	IncludePaths []string `yaml:"include_paths,omitempty" validate:"required"`
-	ExcludePaths []string `yaml:"exclude_paths,omitempty"`
+type LoggingReceiver interface {
+	component
 }
 
-type LoggingReceiverSyslog struct {
-	TransportProtocol string `yaml:"transport_protocol,omitempty"` // one of "tcp" or "udp"
-	ListenHost        string `yaml:"listen_host,omitempty" validate:"required"`
-	ListenPort        uint16 `yaml:"listen_port,omitempty" validate:"required"`
+var loggingReceiverTypes = &componentTypeRegistry{
+	Subagent: "logging", Kind: "receiver",
 }
 
-type LoggingReceiverWinevtlog struct {
-	Channels []string `yaml:"channels,omitempty,flow" validate:"required"`
+// Wrapper type to store the unmarshaled YAML value.
+type loggingReceiverWrapper struct {
+	inner interface{}
 }
 
-type LoggingReceiver struct {
-	configComponent `yaml:",inline"`
+func (l *loggingReceiverWrapper) UnmarshalYAML(ctx context.Context, unmarshal func(interface{}) error) error {
+	return loggingReceiverTypes.unmarshalComponentYaml(ctx, &l.inner, unmarshal)
+}
 
-	LoggingReceiverSyslog    `yaml:",inline"` // Type "syslog"
-	LoggingReceiverFiles     `yaml:",inline"` // Type "files"
-	LoggingReceiverWinevtlog `yaml:",inline"` // Type "windows_event_log"
+func (m *loggingReceiverMap) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	// Unmarshal into a temporary map to capture types.
+	tm := map[string]loggingReceiverWrapper{}
+	if err := unmarshal(&tm); err != nil {
+		return err
+	}
+	// Unwrap the structs.
+	*m = loggingReceiverMap{}
+	for k, r := range tm {
+		(*m)[k] = r.inner.(LoggingReceiver)
+	}
+	return nil
+}
+
+type LoggingProcessor interface {
+	component
+	GetField() string
 }
 
-type LoggingProcessorParseJson struct {
+type LoggingProcessorParseShared struct {
 	Field      string `yaml:"field,omitempty"`       // default to "message"
 	TimeKey    string `yaml:"time_key,omitempty"`    // by default does not parse timestamp
 	TimeFormat string `yaml:"time_format,omitempty"` // must be provided if time_key is present
 }
 
-type LoggingProcessorParseRegex struct {
-	Regex string `yaml:"regex,omitempty" validate:"required"`
+func (p LoggingProcessorParseShared) GetField() string {
```

Comment:
Go style: Avoid starting method names with Get.

Reply:
This should go away in the subsequent PR reworking the config, and we can't name this `Field` because that's already taken, so I'm not going to attempt to fix this.